File: apps/hostd/cmd/guest-agent/stream.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"os/exec"
	"strconv"
	"sync"

	"github.com/coder/websocket"
)

// wsConn is the subset of the websocket connection the frame writer needs,
// kept as an interface so the framing logic is testable without a socket.
type wsConn interface {
	Write(ctx context.Context, typ websocket.MessageType, p []byte) error
}

const msgBinary = websocket.MessageBinary

// handleExecStream runs a command and streams its output frame by frame.
//
// This is the path a long-running agent command takes: output can be megabytes
// and the run can last minutes, so a buffered exec is unusable for it.
//
// Query parameters: repeated `cmd` (argv), `dir`, repeated `env` as K=V,
// `user`, and `stdin=true` to opt into forwarding client messages to the
// process. stdin is off by default -- most callers never write, and a process
// holding an open stdin it never reads can hang.
//
// Frames server to client are 1 stdout, 2 stderr, then the verdict: a text
// {"type":"exit","exit_code":n} and, after it, 3 exit. Client to server they
// are 0 stdin and 4 stdin EOF, and they are read only when the stream opted
// into stdin: with stdin=false nothing is read from the socket at all, so a 0
// frame sent anyway is ignored rather than an error.
//
// tty=true runs the command on a pseudo-terminal instead of three pipes, with
// `rows` and `cols` giving the initial window (24 by 80 by default, each
// 1..65535). It changes four things and nothing else:
//
//   - a PTY merges the two output streams, so everything the command writes
//     arrives as frame 1 and frame 2 is NEVER sent;
//   - a tty implies stdin, so client frames are read whether or not stdin=true
//     was asked for (hostd refuses the contradictory stdin=false up front);
//   - frame 4 writes EOT (0x04) to the terminal rather than closing anything,
//     because a terminal has no separate stdin to close, and the session stays
//     open: what EOT means is the shell's decision, not this agent's;
//   - a TEXT message {"type":"resize","cols":N,"rows":N} resizes the window.
//
// The exit verdict is unchanged, so a client that only reads frames needs no
// tty-specific code to learn how the command ended.
func handleExecStream(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()
	argv := q["cmd"]
	if len(argv) == 0 {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "cmd is required"})
		return
	}

	conn, err := websocket.Accept(w, r, &websocket.AcceptOptions{
		// The connection is authenticated by token before it gets here, and
		// the client is an SDK rather than a browser, so Origin checking adds
		// nothing.
		InsecureSkipVerify: true,
	})
	if err != nil {
		log.Printf("guest-agent: ws accept: %v", err)
		return
	}
	defer conn.CloseNow()

	// After the upgrade, so a bad window size is reported as a close code the
	// client is already listening for rather than an HTTP status it can no
	// longer see.
	tty := q.Get("tty") == "true"
	rows, rowsOK := winDim(q.Get("rows"), 24)
	cols, colsOK := winDim(q.Get("cols"), 80)
	if !rowsOK || !colsOK {
		_ = conn.Close(websocket.StatusPolicyViolation, "rows and cols must be 1..65535")
		return
	}

	// Detached from the HTTP request context: the command should outlive the
	// handler's own lifetime bookkeeping and end only when it exits or the
	// socket closes.
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	cmd := exec.CommandContext(ctx, argv[0], argv[1:]...)
	if err := prepareCommand(cmd, q.Get("user"), q.Get("dir"), nil); err != nil {
		_ = conn.Close(websocket.StatusInternalError, err.Error())
		return
	}
	// After prepareCommand, so the caller's env overrides the account's
	// defaults -- for duplicate keys the last entry wins.
	cmd.Env = append(cmd.Env, q["env"]...)

	fw := &frameWriter{conn: conn, ctx: ctx}

	var (
		stdin io.WriteCloser
		ptmx  *os.File
		pumps sync.WaitGroup
	)

	if tty {
		// One pump, not two: the PTY master IS both output streams, so a
		// second pump would have nothing to read and frame 2 would never be
		// sent anyway. It is stdin as well -- a terminal has one device.
		if ptmx, err = startPTY(cmd, rows, cols); err != nil {
			// 127 is the shell's convention for "could not run it", and is
			// distinguishable from any exit code the command itself could
			// return.
			writeExit(fw, 127)
			_ = conn.Close(websocket.StatusNormalClosure, "start failed")
			return
		}
		defer ptmx.Close()
		pumps.Add(1)
		go pump(fw, frameStdout, ptmx, pumps.Done)
	} else {
		stdout, err := cmd.StdoutPipe()
		if err != nil {
			_ = conn.Close(websocket.StatusInternalError, "stdout pipe")
			return
		}
		stderr, err := cmd.StderrPipe()
		if err != nil {
			_ = conn.Close(websocket.StatusInternalError, "stderr pipe")
			return
		}

		if q.Get("stdin") == "true" {
			if stdin, err = cmd.StdinPipe(); err != nil {
				_ = conn.Close(websocket.StatusInternalError, "stdin pipe")
				return
			}
		}

		if err := cmd.Start(); err != nil {
			writeExit(fw, 127)
			_ = conn.Close(websocket.StatusNormalClosure, "start failed")
			return
		}

		pumps.Add(2)
		go pump(fw, frameStdout, stdout, pumps.Done)
		go pump(fw, frameStderr, stderr, pumps.Done)
	}
	defer trackPID(cmd.Process)()

	// A tty implies stdin: there is no way to type into a terminal otherwise,
	// and hostd refuses tty=true with stdin=false before the machine is woken.
	if stdin != nil || ptmx != nil {
		go func() {
			if stdin != nil {
				defer stdin.Close()
			}
			for {
				typ, data, err := conn.Read(ctx)
				if err != nil {
					// A pipe stream ends itself: the deferred stdin.Close
					// delivers EOF to the process. A terminal has no such end,
					// and an interactive shell writes nothing while it waits
					// for input, so the output pump would park on the master
					// forever and the handler with it -- one orphaned shell,
					// two goroutines and a PTY per closed browser tab. The
					// cancel is what the /terminal handler has always done.
					if ptmx != nil {
						cancel()
					}
					return
				}
				// Text frames carry control messages. Only resize is
				// implemented, and only a tty has a window to resize, so
				// without one this is still the drop it always was.
				if typ == websocket.MessageText {
					var ctl struct {
						Type string `json:"type"`
						Cols uint16 `json:"cols"`
						Rows uint16 `json:"rows"`
					}
					if json.Unmarshal(data, &ctl) == nil && ctl.Type == "resize" {
						resizePTY(ptmx, ctl.Cols, ctl.Rows)
					}
					continue
				}
				// The sprites frame protocol, client to server: byte 0 is the
				// stream id. An empty frame and any id this agent does not know
				// are ignored, which means a legacy raw-stdin client is refused
				// rather than guessed at.
				if typ != websocket.MessageBinary || len(data) == 0 {
					continue
				}
				switch data[0] {
				case frameStdin:
					if _, err := writeStdin(stdin, ptmx, data[1:]); err != nil {
						return
					}
				case frameStdinEOF:
					if ptmx != nil {
						// A terminal has no write end to close, so end-of-file
						// is the EOT character the line discipline turns into
						// one for whatever is reading. The session stays open:
						// what EOT means there is the shell's decision.
						_, _ = ptmx.Write([]byte{4})
						continue
					}
					return // the deferred Close delivers EOF to the process
				}
			}
		}()
	}

	// Both pumps must drain before the exit frame goes out, or a client can
	// see the exit code before the output that preceded it.
	pumps.Wait()
	writeExit(fw, exitCodeOf(cmd.Wait()))
	_ = conn.Close(websocket.StatusNormalClosure, "")
}

// writeExit sends the verdict twice: as text, and then as the binary frame.
//
// The order is the whole point. Websocket frames are ordered, and both SDKs
// act on whichever verdict reaches them first -- they close the socket and
// return -- so a text frame written AFTER the binary one is a frame nothing
// can ever receive. It went out second for a while, which made its documented
// purpose unachievable.
//
// That purpose is real: the binary frame carries the code in one byte, and a
// command killed by a signal has an ExitCode of -1, which byte() reports as
// 255 -- indistinguishable from a command that genuinely exited 255. The text
// form carries the code as an integer, so it has to be the one that lands
// first.
//
// The binary frame still follows, because it is the sprites byte protocol and
// a client that reads only binary frames is the whole reason to keep it.
func writeExit(fw *frameWriter, code int) {
	_ = fw.writeText(fmt.Appendf(nil, `{"type":"exit","exit_code":%d}`, code))
	_ = fw.write(frameExit, []byte{byte(code)})
}

// winDim parses a rows or cols query value. An absent value takes the default;
// anything that is not 1..65535 is refused rather than clamped, because a
// client that computed a window size wrongly should learn that at the
// handshake and not from a terminal that silently drew itself 80 columns wide.
func winDim(raw string, def uint16) (uint16, bool) {
	if raw == "" {
		return def, true
	}
	n, err := strconv.ParseUint(raw, 10, 16)
	if err != nil || n == 0 {
		return 0, false
	}
	return uint16(n), true
}

// writeStdin sends one client chunk to whichever input the stream opened.
func writeStdin(stdin io.WriteCloser, ptmx *os.File, chunk []byte) (int, error) {
	if ptmx != nil {
		return ptmx.Write(chunk)
	}
	return stdin.Write(chunk)
}

File: sdks/python/src/pilots/stream.py
"""The streaming exec.

The wire format is sprites' byte protocol, so an existing sprites client
drops in unchanged: every binary frame's first byte is the stream id --
1 stdout, 2 stderr, 3 exit with the code in ``payload[0]`` -- and the two the
client sends are 0 for a stdin chunk and 4 for stdin EOF.

Two behaviours here are load-bearing rather than stylistic:

* ``stdin`` defaults to False. A process holding an open stdin it never reads
  hangs, and an agent run under ``claude -p`` is exactly such a process.
* a close with no exit frame is an ERROR, never a silent code 0. The guest
  agent drains both output pumps before it writes the exit frame and
  websocket frames are ordered, so an exit frame means the output that
  preceded it has already arrived. A socket that dropped instead means nobody
  knows what the command did.

``tty`` is a mode on this same stream, not a second protocol: a PTY merges the
two output streams onto ``stdout``, stdin is always read, ``end_stdin()``
sends EOT rather than closing anything, and ``resize()`` becomes callable.
"""

from __future__ import annotations

import json
import queue
import threading
from collections.abc import Iterator
from typing import Any
from urllib.parse import urlencode

from .errors import PilotsError
from .types import FrameExit, FrameStderr, FrameStdin, FrameStdinEOF, FrameStdout

_EOF = object()


def build_exec_url(
    base_url: str,
    path: str,
    argv: list[str],
    *,
    cwd: str | None = None,
    env: dict[str, str] | None = None,
    user: str | None = None,
    stdin: bool = False,
    tty: bool = False,
    rows: int | None = None,
    cols: int | None = None,
    org: str | None = None,
) -> str:
    """Builds the exec-stream URL, with the query names sprites uses."""
    if base_url.startswith("http"):
        base_url = "ws" + base_url[len("http") :]
    params: list[tuple[str, str]] = []
    # The org narrowing reaches THIS route too; every other call applies it
    # in Http, and this one builds its own URL.
    if org:
        params.append(("org", org))
    for arg in argv:
        params.append(("cmd", arg))
    if argv:
        params.append(("path", argv[0]))
    if cwd:
        params.append(("dir", cwd))
    for k, v in (env or {}).items():
        params.append(("env", f"{k}={v}"))
    if user:
        params.append(("user", user))
    # Always present, never inferred: the default is the thing most likely to
    # be wrong by omission. A tty forces it on; hostd refuses the pair
    # tty=true&stdin=false with a 400.
    params.append(("stdin", "true" if (tty or stdin) else "false"))
    if tty:
        params.append(("tty", "true"))
        if rows is not None:
            params.append(("rows", str(rows)))
        if cols is not None:
            params.append(("cols", str(cols)))
    return f"{base_url}{path}?{urlencode(params)}"


class _Pipe:
    """A readable end fed by the reader thread: iterate it for chunks, or
    ``read()`` it whole once the stream has exited."""

    def __init__(self) -> None:
        self._q: queue.Queue[Any] = queue.Queue()
        self._closed = False

    def _push(self, chunk: bytes) -> None:
        if not self._closed:
            self._q.put(chunk)

    def _close(self) -> None:
        if not self._closed:
            self._closed = True
            self._q.put(_EOF)

    def __iter__(self) -> Iterator[bytes]:
        while True:
            item = self._q.get()
            if item is _EOF:
                self._q.put(_EOF)
                return
            yield item

    def read(self) -> bytes:
        """Everything until EOF, in one bytes object."""
        return b"".join(self)


class ExecStream:
    """A running command.

    ``stdout`` and ``stderr`` are queues fed by a reader thread. A WebSocket
    cannot be paused, so those queues are the only boundary: a stream nobody
    reads grows without limit. For output nobody intends to read, use the
    buffered ``machines.exec``.
    """

    def __init__(self, url: str, api_key: str, *, stdin: bool, tty: bool, open_timeout: float = 30.0) -> None:
        from websockets.sync.client import connect

        self.stdout = _Pipe()
        self.stderr = _Pipe()
        self._stdin_enabled = stdin or tty
        self._tty = tty
        self._code: int | None = None
        self._exited = False
        self._error: Exception | None = None
        self._done = threading.Event()
        self._lock = threading.Lock()
        try:
            # The key travels as a subprotocol rather than a header: one code
            # path shared with browsers, and hostd accepts either form.
            self._ws = connect(
                url,
                subprotocols=[f"authorization.bearer.{api_key}"],  # type: ignore[list-item]
                open_timeout=open_timeout,
                max_size=None,
            )
        except Exception as cause:
            raise PilotsError(f"exec stream could not connect: {cause}") from cause
        self._reader = threading.Thread(target=self._pump, name="pilots-exec-stream", daemon=True)
        self._reader.start()

    # -- the reader -----------------------------------------------------------

    def _pump(self) -> None:
        try:
            for message in self._ws:
                if isinstance(message, str):
                    self._on_text(message)
                elif message:
                    self._on_frame(bytes(message))
        except Exception as err:  # the socket dropped
            self._fail(PilotsError(f"exec stream failed: {err}"))
        finally:
            self._end_pipes()
            if not self._exited:
                self._fail(PilotsError("stream closed before exit"))
            self._done.set()

    def _on_frame(self, frame: bytes) -> None:
        kind, payload = frame[0], frame[1:]
        if kind == FrameStdout:
            if not self._exited:
                self.stdout._push(payload)
        elif kind == FrameStderr:
            if not self._exited:
                self.stderr._push(payload)
        elif kind == FrameExit:
            self._finish(payload[0] if payload else 0)
        # An id this version does not know is not a reason to fail.

    def _on_text(self, data: str) -> None:
        # hostd sends a text {"type":"exit","exit_code":n} BEFORE the binary
        # exit frame, because the binary one carries the code in one byte and
        # a signal death (-1) cannot be told from 255 there. Whichever arrives
        # first decides; the other is ignored.
        try:
            parsed = json.loads(data)
        except ValueError:
            return
        if isinstance(parsed, dict) and parsed.get("type") == "exit" and not self._exited:
            self._finish(int(parsed.get("exit_code", 0)))

    def _finish(self, code: int) -> None:
        with self._lock:
            if self._exited:
                return
            self._exited = True
            self._code = code
        self._end_pipes()
        try:
            self._ws.close()
        except Exception:
            pass

    def _end_pipes(self) -> None:
        self.stdout._close()
        self.stderr._close()

    def _fail(self, err: Exception) -> None:
        with self._lock:
            if self._error is None and not self._exited:
                self._error = err

    # -- the caller's side ------------------------------------------------------

    @property
    def exit_code(self) -> int | None:
        """The exit code, or None until the exit frame arrives."""
        return self._code

    def wait(self, timeout: float | None = None) -> int:
        """Blocks until exit and returns the code; raises on a failure or a
        close before exit."""
        if not self._done.wait(timeout):
            raise PilotsError("timed out waiting for the command to exit")
        if self._error is not None:
            raise self._error
        assert self._code is not None
        return self._code

    def output(self, timeout: float | None = None) -> tuple[bytes, bytes, int]:
        """Waits for the exit, then drains both pipes: ``(stdout, stderr, exit_code)``.

        The WAIT comes first, and the order is the whole reason ``timeout``
        means anything: a pipe is only closed by the exit frame or a dropped
        socket, so draining first blocks forever on a command that is alive and
        silent -- and ``timeout`` would bound nothing. Once the exit has
        arrived, both pipes are already at EOF and the reads return at once.
        """
        code = self.wait(timeout)
        out_box: list[bytes] = []
        err_box: list[bytes] = []
        t = threading.Thread(target=lambda: err_box.append(self.stderr.read()), daemon=True)
        t.start()
        out_box.append(self.stdout.read())
        t.join()
        return out_box[0], err_box[0], code

    def write_stdin(self, chunk: bytes | str) -> None:
        """Sends one stdin chunk (frame 0). Raises unless the stream opted into stdin."""
        if not self._stdin_enabled:
            raise PilotsError("this stream was opened with stdin=False")
        data = chunk.encode() if isinstance(chunk, str) else bytes(chunk)
        self._ws.send(bytes([FrameStdin]) + data)

    def end_stdin(self) -> None:
        """Closes the process's stdin (frame 4). Under ``tty`` this sends EOT
        to the terminal instead, and the session stays open."""
        if not self._stdin_enabled:
            raise PilotsError("this stream was opened with stdin=False")
        self._ws.send(bytes([FrameStdinEOF]))

    def resize(self, cols: int, rows: int) -> None:
        """Resizes the terminal. Raises unless the stream was opened with ``tty``."""
        if not self._tty:
            raise PilotsError("this stream was opened without tty")
        self._ws.send(json.dumps({"type": "resize", "cols": cols, "rows": rows}))

    def kill(self) -> None:
        """Closes the socket. The agent's context cancel kills the process."""
        try:
            self._ws.close()
        except Exception:
            pass

    close = kill

    def __enter__(self) -> ExecStream:
        return self

    def __exit__(self, *exc: object) -> None:
        self.kill()
